### PR TITLE
refactor: remove dependency for TileFormatter and determine tile form…

### PR DESCRIPTION
**Pull request by shimoncohen**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✖                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✖                                                                        |
| Chore            | ✔                                                                       |

### CODE REVIEW

File: MergerLogic/Clients/GpkgClient.cs [hunk 1]
```diff
@@ -140,18 +139,11 @@ public override Tile GetTile(int z, int x, int y)
...
-                        return null;
-                    }
-
-                    tile = new Tile(z, x, y, blob);
+                    blob = (byte[])command.ExecuteScalar();
                 }
             }
 
-            return tile;
+            return this.CreateTile(z, x, y, blob);
```
Comment (asafMasa) [inline MergerLogic/Clients/GpkgClient.cs:146]: dont return a method (separate to a different line) - it's a bad practice (hard to follow logs and debug later)

File: MergerLogic/Batching/Tile.cs [hunk 2]
```diff
@@ -29,22 +29,22 @@ public int Y
...
 
-        public Tile(Coord cords, byte[] data, TileFormat? format = null)
+        public Tile(Coord cords, byte[] data)
         {
             this.Z = cords.Z;
             this.X = cords.X;
             this.Y = cords.Y;
             this._data = data;
-            this.Format = format;
+            this.Format = ImageFormatter.GetTileFormat(data);
```
Comment (asafMasa) [inline MergerLogic/Batching/Tile.cs:47]: throw a validation error if  data is set to null

File: MergerLogic/Batching/Tile.cs [hunk 3]
```diff
@@ -29,22 +29,22 @@ public int Y
...
 
-        public Tile(int z, int x, int y, byte[] data, TileFormat? format = null)
+        public Tile(int z, int x, int y, byte[] data)
         {
             this.Z = z;
             this.X = x;
             this.Y = y;
             this._data = data;
-            this.Format = format;
+            this.Format = ImageFormatter.GetTileFormat(data);
```
Comment (asafMasa) [inline MergerLogic/Batching/Tile.cs:38]: throw a validation error if  data is set to null

File: MergerLogic/Clients/GpkgClient.cs [hunk 4]
```diff
@@ -239,7 +231,7 @@ public List<Tile> GetBatch(int batchSize, long offset)
                             var y = reader.GetInt32(2);
                             var blob = (byte[])reader["tile_data"];
 
-                            Tile tile = new Tile(z, x, y, blob);
+                            Tile tile = this.CreateTile(z, x, y, blob)!;
```
Comment (asafMasa) [inline MergerLogic/Clients/GpkgClient.cs:234]: why did you put "!" at this line?

File: MergerLogic/Clients/GpkgClient.cs [hunk 5]
```diff
@@ -298,7 +290,7 @@ public List<Tile> GetBatch(int batchSize, long offset)
                         var y = reader.GetInt32(2);
                         var blob = (byte[])reader["tile_data"];
 
-                        lastTile = new Tile(z, x, y, blob);
+                        lastTile = this.CreateTile(z, x, y, blob)!;
```
Comment (asafMasa) [inline MergerLogic/Clients/GpkgClient.cs:293]: why did you put "!" at this line?

File: MergerLogic/ImageProcessing/ImageFormatter.cs [hunk 6]
```diff
@@ -10,40 +9,24 @@ public enum TileFormat
...
-                }
-            }
-
-            return tile;
-        }
-
-        public byte[] ConvertToFormat(byte[] tile, TileFormat format)
-        {
-            var currentFormat = this.GetTileFormat(tile);
+            var currentFormat = GetTileFormat(tile);
```
Comment (asafMasa) [inline MergerLogic/ImageProcessing/ImageFormatter.cs:16]: with static methods don't we need to call it with the class Name? 
ImageFormatter.GetTileFormat(tile) ?

**Review by asafMasa — CHANGES_REQUESTED** [review]
please mind my remarks

File: MergerLogic/Batching/Tile.cs [hunk 7]
```diff
@@ -25,26 +26,26 @@ public int Y
...
 
-        public Tile(int z, int x, int y, byte[] data, TileFormat? format = null)
+        public Tile(int z, int x, int y, byte[] data)
         {
             this.Z = z;
             this.X = x;
             this.Y = y;
             this._data = data;
-            this.Format = format;
+            this.Format = ImageFormatter.GetTileFormat(data) ?? throw new ValidationException($"Cannot create tile {this}, data is in invalid format");
```
Comment (asafMasa) [inline MergerLogic/Batching/Tile.cs:39]: please put it on "this._data = data;" - as ImageFormatter.GetTileFormat will throw on null reference

File: MergerLogic/Batching/Tile.cs [hunk 8]
```diff
@@ -25,26 +26,26 @@ public int Y
...
 
-        public Tile(Coord cords, byte[] data, TileFormat? format = null)
+        public Tile(Coord cords, byte[] data)
         {
             this.Z = cords.Z;
             this.X = cords.X;
             this.Y = cords.Y;
             this._data = data;
-            this.Format = format;
+            this.Format = ImageFormatter.GetTileFormat(data) ?? throw new ValidationException($"Cannot create tile {this}, data is in invalid format");
```
Comment (asafMasa) [inline MergerLogic/Batching/Tile.cs:48]: please put it on "this._data = data;" - as ImageFormatter.GetTileFormat will throw on null reference

File: MergerLogicUnitTests/Batching/TileTest.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+            byte[] data = { 0x43, 0x44, 0x30, 0x30, 0x31 };
+            Assert.ThrowsException<ValidationException>(() => new Tile(0, 0, 0, data));
+            this._repository.VerifyAll();
+        }
+
+        [TestMethod]
+        [TestCategory("CreateTile")]
+        [DataRow(new byte[] { 0xFF, 0xD8, 0xFF, 0xDB })]
+        [DataRow(new byte[] { 0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A })]
+        public void CreateTile(byte[] data)
```
Comment (asafMasa) [inline MergerLogicUnitTests/Batching/TileTest.cs:38]: please add null and check if validation error is thrown

**Review by asafMasa — CHANGES_REQUESTED** [review]
please mind my change requests

File: MergerLogic/Batching/Tile.cs (hunk 7 above)
Comment (shimoncohen) [reply MergerLogic/Batching/Tile.cs:39]: The issue is when trying to extract the format, not when assigning data.
You shouldn't get `null` data.

File: MergerLogic/Batching/Tile.cs (hunk 8 above)
Comment (shimoncohen) [reply MergerLogic/Batching/Tile.cs:48]: The issue is when trying to extract the format, not when assigning data.
You shouldn't get `null` data.

File: MergerLogicUnitTests/Batching/TileTest.cs (hunk 9 above)
Comment (shimoncohen) [reply MergerLogicUnitTests/Batching/TileTest.cs:38]: This isn't the correct place, added in another test.

**Review by asafMasa — APPROVED** [review]
